cryptography/src/handshake.rs

```diff
@@ -0,0 +1,246 @@
+use chacha20poly1305::{aead::AeadMut, ChaCha20Poly1305, KeyInit};
+use commonware_codec::Encode;
+use rand_core::CryptoRngCore;
+
+use crate::{
+    transcript::{Summary, Transcript},
+    PublicKey, Signer, Verifier,
+};
+
+const NAMESPACE: &'static [u8] = b"commonware/handshake";
+
+pub struct Msg1<S> {
+    time_ms: u64,
+    epk: x25519_dalek::PublicKey,
+    sig: S,
+}
+
+pub struct Msg2<S> {
+    time_ms: u64,
+    epk: x25519_dalek::PublicKey,
+    sig: S,
+    confirmation: Summary,
+}
+
+pub struct Msg3 {
+    confirmation: Summary,
+}
+
+pub struct DialState<P> {
+    esk: x25519_dalek::EphemeralSecret,
+    peer_identity: P,
+    transcript: Transcript,
+}
+
+pub struct ListenState {
+    confirmation: Summary,
+    send: SendCipher,
+    recv: RecvCipher,
+}
+
+struct CounterNonce {
+    inner: u128,
+}
+
+impl CounterNonce {
+    pub fn new() -> Self {
+        Self { inner: 0 }
+    }
+
+    pub fn inc(&mut self) -> [u8; 16] {
+        if self.inner >= 1 << 96 {
+            panic!("overflowed nonce");
+        }
+        let out = self.inner.to_le_bytes();
+        self.inner += 1;
+        out
+    }
+}
+
+pub struct SendCipher {
+    nonce: CounterNonce,
+    inner: ChaCha20Poly1305,
+}
+
+impl SendCipher {
+    fn new(mut rng: impl CryptoRngCore) -> Self {
+        let mut key = [0u8; 32];
+        rng.fill_bytes(&mut key[..]);
+        Self {
+            nonce: CounterNonce::new(),
+            inner: ChaCha20Poly1305::new(&key.into()),
+        }
+    }
+
+    pub fn send(&mut self, data: &[u8]) -> Vec<u8> {
+        self.inner
+            .encrypt((&self.nonce.inc()[..12]).into(), data)
+            .unwrap()
+    }
+}
+
+pub struct RecvCipher {
+    nonce: CounterNonce,
+    inner: ChaCha20Poly1305,
+}
+
+impl RecvCipher {
+    fn new(mut rng: impl CryptoRngCore) -> Self {
+        let mut key = [0u8; 32];
+        rng.fill_bytes(&mut key[..]);
+        Self {
+            nonce: CounterNonce::new(),
+            inner: ChaCha20Poly1305::new(&key.into()),
+        }
+    }
+
+    pub fn recv(&mut self, encrypted_data: &[u8]) -> Vec<u8> {
+        self.inner
+            .decrypt((&self.nonce.inc()[..12]).into(), encrypted_data)
+            .unwrap()
+    }
+}
+
+pub struct Context<S, P> {
+    current_time: u64,
+    my_identity: S,
+    peer_identity: P,
+}
+
+impl<S, P> Context<S, P> {
+    pub fn new(current_time_ms: u64, my_identity: S, peer_identity: P) -> Self {
+        Self {
+            current_time: current_time_ms,
+            my_identity,
+            peer_identity,
+        }
+    }
+}
+
+pub fn dial_start<S: Signer, P: PublicKey>(
```

Comment:
Wonder if it would be better to use something more generic than `dial` (given there aren't really network connections here per se).